File: mobile/ios/Hearth/Cloud.swift
import AuthenticationServices
import CryptoKit
import Foundation
import Security
import UIKit

/// Signing in to Google Drive and Dropbox for backups (docs/decisions/0009-cloud-drive-buttons.md).
///
/// It touches nothing but credentials, and its requests go through Native/Egress.swift, the app's
/// one network gateway: it signs the user in, keeps the long-lived grant in the Keychain, and hands
/// the page short-lived access tokens on request. The backup itself (encrypted when the user set a passphrase) goes from the page through
/// `frontend/src/egress/egress.ts`, so the one place that decides what leaves the device stays one.
///
/// Both providers use OAuth with PKCE in `ASWebAuthenticationSession`, the system's sign-in sheet:
/// no client secret exists anywhere, only a public client id (Google's iOS client) or app key
/// (Dropbox), set at build time from the gitignored .env. Drive asks for the full `drive` scope so
/// the user can choose any folder (Hearth reads and writes only there); Dropbox keeps to an app folder.
final class Cloud: NSObject, ASWebAuthenticationPresentationContextProviding {
    /// The answer to a sign-in or token request: a value, or an error message.
    typealias Reply = (Any?, String?) -> Void

    struct Provider {
        let authorize: URL
        let token: URL
        let revoke: URL
        let clientID: String
        let redirect: String
        let scheme: String
        let extra: [URLQueryItem]
    }

    /// Access tokens already handed out, so a backup every few seconds does not refresh each time.
    /// Main thread only.
    private var tokens: [String: (token: String, until: Date)] = [:]
    private var session: ASWebAuthenticationSession?

    static func provider(_ name: String, bundle: Bundle = .main) -> Provider? {
        switch name {
        case "google":
            // An iOS OAuth client: `<id>.apps.googleusercontent.com`, whose reversed form is the
            // redirect scheme Google accepts for it.
            guard let id = setting("HearthGoogleClientID", bundle), id.hasSuffix(".apps.googleusercontent.com")
            else { return nil }
            let scheme = "com.googleusercontent.apps." + id.dropLast(".apps.googleusercontent.com".count)
            return Provider(
                authorize: URL(string: "https://accounts.google.com/o/oauth2/v2/auth")!,
                token: URL(string: "https://oauth2.googleapis.com/token")!,
                revoke: URL(string: "https://oauth2.googleapis.com/revoke")!,
                clientID: id,
                redirect: "\(scheme):/oauth2redirect",
                scheme: scheme,
                extra: [URLQueryItem(name: "scope", value: "openid email https://www.googleapis.com/auth/drive")]
            )
        case "dropbox":
            guard let key = setting("HearthDropboxAppKey", bundle) else { return nil }
            return Provider(
                authorize: URL(string: "https://www.dropbox.com/oauth2/authorize")!,
                token: URL(string: "https://api.dropboxapi.com/oauth2/token")!,
                revoke: URL(string: "https://api.dropboxapi.com/2/auth/token/revoke")!,
                clientID: key,
                redirect: "db-\(key)://2/token",
                scheme: "db-\(key)",
                extra: [URLQueryItem(name: "token_access_type", value: "offline")]
            )
        default:
            return nil
        }
    }

    /// An Info.plist value filled from a build setting; empty or unexpanded means "not set up".
    private static func setting(_ key: String, _ bundle: Bundle) -> String? {
        guard let value = bundle.object(forInfoDictionaryKey: key) as? String,
              !value.isEmpty, !value.hasPrefix("$(")
        else { return nil }
        return value
    }

    // MARK: - The page's calls (main thread)

    func signIn(_ name: String, reply: @escaping Reply) {
        guard let provider = Self.provider(name) else {
            reply(nil, "\(name) is not set up for this build of Hearth")
            return
        }
        guard session == nil else {
            reply(nil, "a sign-in is already open")
            return
        }
        let verifier = Self.randomToken(48)
        let state = Self.randomToken(16)
        var url = URLComponents(url: provider.authorize, resolvingAgainstBaseURL: false)!
        url.queryItems = [
            URLQueryItem(name: "client_id", value: provider.clientID),
            URLQueryItem(name: "response_type", value: "code"),
            URLQueryItem(name: "redirect_uri", value: provider.redirect),
            URLQueryItem(name: "code_challenge", value: Self.challenge(verifier)),
            URLQueryItem(name: "code_challenge_method", value: "S256"),
            URLQueryItem(name: "state", value: state),
        ] + provider.extra
        let sheet = ASWebAuthenticationSession(url: url.url!, callbackURLScheme: provider.scheme) { [weak self] callback, error in
            DispatchQueue.main.async {
                self?.session = nil
                if let error = error as? ASWebAuthenticationSessionError, error.code == .canceledLogin {
                    reply(NSNull(), nil)
                    return
                }
                let items = callback.flatMap { URLComponents(url: $0, resolvingAgainstBaseURL: false)?.queryItems } ?? []
                let value = { (key: String) in items.first { $0.name == key }?.value }
                guard value("state") == state, let code = value("code") else {
                    // Denied on the provider's screen, or not the answer to our request.
                    reply(NSNull(), nil)
                    return
                }
                self?.exchange(name, provider, code: code, verifier: verifier, reply: reply)
            }
        }
        sheet.presentationContextProvider = self
        session = sheet
        if !sheet.start() {
            session = nil
            reply(nil, "could not open the sign-in sheet")
        }
    }

    /// A fresh access token, or NSNull when the user has to sign in again.
    func token(_ name: String, account: String, stale: String?, reply: @escaping Reply) {
        let key = "\(name):\(account)"
        if let cached = tokens[key], cached.token != stale, cached.until > Date() {
            reply(cached.token, nil)
            return
        }
        guard let provider = Self.provider(name), let refresh = Keychain.read(key) else {
            reply(NSNull(), nil)
            return
        }
        post(provider.token, form: [
            "grant_type": "refresh_token",
            "refresh_token": refresh,
            "client_id": provider.clientID,
        ]) { [weak self] result in
            switch result {
            case .success(let body):
                guard let access = body["access_token"] as? String else {
                    reply(nil, "the provider sent no token")
                    return
                }
                self?.remember(key, access, body["expires_in"])
                reply(access, nil)
            case .failure(let error as HTTPError) where error.status == 400:
                // The grant is gone (revoked, or the app unlinked): sign in again.
                Keychain.delete(key)
                reply(NSNull(), nil)
            case .failure(let error):
                reply(nil, error.localizedDescription)
            }
        }
    }

    func signOut(_ name: String, account: String, reply: @escaping Reply) {
        let key = "\(name):\(account)"
        tokens[key] = nil
        if let provider = Self.provider(name), let refresh = Keychain.read(key) {
            // Revoked at the provider too, so the grant is not just forgotten but dead. Best effort.
            if name == "google" {
                post(provider.revoke, form: ["token": refresh]) { _ in }
            } else {
                token(name, account: account, stale: nil) { access, _ in
                    guard let access = access as? String else { return }
                    Egress.request("POST", provider.revoke, headers: ["Authorization": "Bearer \(access)"]) { _ in }
                }
            }
        }
        Keychain.delete(key)
        reply(NSNull(), nil)
    }

    // MARK: - Tokens

    private func exchange(_ name: String, _ provider: Provider, code: String, verifier: String, reply: @escaping Reply) {
        post(provider.token, form: [
            "grant_type": "authorization_code",
            "code": code,
            "code_verifier": verifier,
            "client_id": provider.clientID,
            "redirect_uri": provider.redirect,
        ]) { [weak self] result in
            guard case .success(let body) = result,
                  let access = body["access_token"] as? String,
                  let refresh = body["refresh_token"] as? String
            else {
                if case .failure(let error) = result { reply(nil, error.localizedDescription) }
                else { reply(nil, "the provider sent no token") }
                return
            }
            let (account, label) = Self.identity(name, body)
            Keychain.write("\(name):\(account)", refresh)
            self?.remember("\(name):\(account)", access, body["expires_in"])
            reply(["account": account, "label": label, "token": access] as [String: Any], nil)
        }
    }

    /// Who signed in: Dropbox says so in the token answer, Google in the ID token's claims (read
    /// for display only; it came straight from Google over TLS).
    static func identity(_ name: String, _ body: [String: Any]) -> (account: String, label: String) {
        if name == "dropbox" {
            let id = body["account_id"] as? String ?? "dropbox"
            return (id, "Dropbox")
        }
        let claims = (body["id_token"] as? String).flatMap(jwtClaims) ?? [:]
        let email = claims["email"] as? String ?? ""
        let subject = claims["sub"] as? String ?? email
        return (email.isEmpty ? subject : email, email.isEmpty ? "Google" : email)
    }

    static func jwtClaims(_ jwt: String) -> [String: Any]? {
        let parts = jwt.split(separator: ".")
        guard parts.count >= 2 else { return nil }
        var payload = parts[1].replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
        payload += String(repeating: "=", count: (4 - payload.count % 4) % 4)
        guard let data = Data(base64Encoded: payload) else { return nil }
        return (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
    }

    private func remember(_ key: String, _ token: String, _ expiresIn: Any?) {
        // A minute early, so a token never expires between being handed out and being used.
        let seconds = (expiresIn as? NSNumber)?.doubleValue ?? 3600
        tokens[key] = (token, Date().addingTimeInterval(seconds - 60))
    }

    // MARK: - Plumbing

    struct HTTPError: LocalizedError {
        let status: Int
        var errorDescription: String? { "the provider answered \(status)" }
    }

    /// A form POST through the app's one network gateway (Native/Egress.swift); the JSON answer of
    /// a 2xx, delivered on the main thread.
    private func post(_ url: URL, form: [String: String], done: @escaping (Result<[String: Any], Error>) -> Void) {
        Egress.request(
            "POST", url,
            headers: ["Content-Type": "application/x-www-form-urlencoded"],
            body: Self.formBody(form).data(using: .utf8),
            timeout: 30
        ) { outcome in
            let result: Result<[String: Any], Error>
            switch outcome {
            case .success(let data):
                result = .success(((try? JSONSerialization.jsonObject(with: data)) as? [String: Any]) ?? [:])
            case .failure(let error):
                if case .http(let status, _)? = error as? Egress.Failure {
                    result = .failure(HTTPError(status: status))
                } else {
                    result = .failure(error)
                }
            }
            DispatchQueue.main.async { done(result) }
        }
    }

    static func formBody(_ fields: [String: String]) -> String {
        var allowed = CharacterSet.alphanumerics
        allowed.insert(charactersIn: "-._~")
        return fields.sorted { $0.key < $1.key }
            .map { key, value in
                "\(key.addingPercentEncoding(withAllowedCharacters: allowed)!)=\(value.addingPercentEncoding(withAllowedCharacters: allowed)!)"
            }
            .joined(separator: "&")
    }

    /// RFC 7636 S256: base64url of the SHA-256 of the verifier, unpadded.
    static func challenge(_ verifier: String) -> String {
        base64url(Data(SHA256.hash(data: Data(verifier.utf8))))
    }

    static func randomToken(_ count: Int) -> String {
        var bytes = [UInt8](repeating: 0, count: count)
        _ = SecRandomCopyBytes(kSecRandomDefault, count, &bytes)
        return base64url(Data(bytes))
    }

    static func base64url(_ data: Data) -> String {
        data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
        let scene = UIApplication.shared.connectedScenes.first { $0.activationState == .foregroundActive }
        return (scene as? UIWindowScene)?.keyWindow ?? ASPresentationAnchor()
    }
}

/// Refresh tokens and the backup passphrase, one generic-password item each, readable only after
/// the first unlock and never migrating to another device through a backup.
enum Keychain {
    private static let service = "hearth.cloud"

    private static func query(_ account: String) -> [String: Any] {
        [kSecClass as String: kSecClassGenericPassword,
         kSecAttrService as String: service,
         kSecAttrAccount as String: account]
    }

    static func read(_ account: String) -> String? {
        var query = query(account)
        query[kSecReturnData as String] = true
        query[kSecMatchLimit as String] = kSecMatchLimitOne
        var item: CFTypeRef?
        guard SecItemCopyMatching(query as CFDictionary, &item) == errSecSuccess,
              let data = item as? Data
        else { return nil }
        return String(data: data, encoding: .utf8)
    }

    static func write(_ account: String, _ value: String) {
        delete(account)
        var item = query(account)
        item[kSecValueData as String] = Data(value.utf8)
        item[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
        SecItemAdd(item as CFDictionary, nil)
    }

    static func delete(_ account: String) {
        SecItemDelete(query(account) as CFDictionary)
    }
}

File: mobile/android/app/src/main/kotlin/com/hearth/Cloud.kt
package com.hearth

import android.accounts.Account
import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Handler
import android.os.Looper
import androidx.activity.ComponentActivity
import androidx.activity.result.IntentSenderRequest
import androidx.activity.result.contract.ActivityResultContracts
import com.google.android.gms.auth.GoogleAuthUtil
import com.google.android.gms.auth.api.identity.AuthorizationRequest
import com.google.android.gms.auth.api.identity.AuthorizationResult
import com.google.android.gms.auth.api.identity.Identity
import com.google.android.gms.common.api.ApiException
import com.google.android.gms.common.api.CommonStatusCodes
import com.google.android.gms.common.api.Scope
import java.io.IOException
import java.net.URLEncoder
import java.security.MessageDigest
import java.security.SecureRandom
import java.util.Base64
import java.util.concurrent.Executors
import org.json.JSONArray
import org.json.JSONObject

/**
 * Signing in to Google Drive and Dropbox for backups (docs/decisions/0009-cloud-drive-buttons.md).
 *
 * This is the shell's only network code, and it touches nothing but credentials: it signs the user
 * in, keeps the long-lived grant, and hands the page short-lived access tokens on request. The
 * backup itself (encrypted when the user set a passphrase) goes from the page through
 * `frontend/src/egress/egress.ts`, so the one place that decides what leaves the device stays one.
 *
 * Google uses Play services' authorization client: no client id or secret in the app, because Google
 * recognises the app by its package name and signing certificate, registered by whoever publishes
 * it. Play services keeps the grant; asking again is silent once the user has agreed. Dropbox uses
 * OAuth with PKCE in the browser, back into the app through its `db-<app key>` scheme; its refresh
 * token is kept in the app's private preferences, beside the database it protects and under the
 * same `allowBackup="false"`.
 *
 * Drive asks for the full `drive` scope, so the user can choose any folder for the backup (a
 * restricted scope: fine for test users, Google verification before a public release); Hearth reads
 * and writes only in the chosen folder. Dropbox keeps to an app folder.
 */
class Cloud(private val activity: ComponentActivity) {
    private val main = Handler(Looper.getMainLooper())
    private val net = Executors.newSingleThreadExecutor()
    private val prefs = activity.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
    private val random = SecureRandom()

    /** Access tokens already handed out, so a backup every few seconds does not refresh each time. */
    private val dropboxTokens = mutableMapOf<String, Pair<String, Long>>()

    private var googleWaiting: Done? = null
    private var dropboxWaiting: DropboxPending? = null

    private val googleConsent =
        activity.registerForActivityResult(ActivityResultContracts.StartIntentSenderForResult()) { result ->
            val done = googleWaiting ?: return@registerForActivityResult
            googleWaiting = null
            try {
                val auth = Identity.getAuthorizationClient(activity).getAuthorizationResultFromIntent(result.data)
                done(Result.success(signedIn(auth)))
            } catch (e: ApiException) {
                // Backing out of Google's screen is an answer, not an error.
                if (e.statusCode == CommonStatusCodes.CANCELED) done(Result.success(JSONObject.NULL))
                else done(Result.failure(googleError(e)))
            }
        }

    /** The providers this build was set up for; the page offers a button for each. */
    fun providers(): JSONArray = JSONArray().apply {
        if (BuildConfig.GOOGLE_DRIVE) put("google")
        if (BuildConfig.DROPBOX_APP_KEY.isNotEmpty()) put("dropbox")
    }

    // ---- the page's calls; all start on the main thread and answer through `done` -------------

    fun signIn(provider: String, done: Done) = when (provider) {
        "google" -> googleSignIn(done)
        "dropbox" -> dropboxSignIn(done)
        else -> done(Result.failure(IllegalArgumentException("unknown provider")))
    }

    /** A fresh access token, or JSON null when the user has to sign in again. */
    fun token(provider: String, account: String, stale: String?, done: Done) = when (provider) {
        "google" -> googleToken(account, stale, done)
        "dropbox" -> background(done) { dropboxToken(account, stale) ?: JSONObject.NULL }
        else -> done(Result.failure(IllegalArgumentException("unknown provider")))
    }

    fun signOut(provider: String, account: String, done: Done) {
        if (provider != "dropbox") return done(Result.success(JSONObject.NULL))
        background(done) {
            // Revoked at Dropbox too, so the grant is not just forgotten but dead.
            dropboxToken(account, null)?.let { runCatching { post(DROPBOX_REVOKE, "null", bearer = it) } }
            prefs.edit().remove(key(account)).apply()
            dropboxTokens.remove(account)
            JSONObject.NULL
        }
    }

    // ---- Google ------------------------------------------------------------------------------

    private fun googleRequest(account: String?) = AuthorizationRequest.Builder()
        .setRequestedScopes(listOf(Scope(DRIVE), Scope("email")))
        .apply { if (!account.isNullOrEmpty()) setAccount(Account(account, GOOGLE_ACCOUNT_TYPE)) }
        .build()

    private fun googleSignIn(done: Done) {
        if (googleWaiting != null) return done(Result.failure(IllegalStateException("a sign-in is already open")))
        Identity.getAuthorizationClient(activity).authorize(googleRequest(null))
            .addOnSuccessListener { auth ->
                val pending = auth.pendingIntent
                if (auth.hasResolution() && pending != null) {
                    googleWaiting = done
                    googleConsent.launch(IntentSenderRequest.Builder(pending.intentSender).build())
                } else {
                    done(Result.success(signedIn(auth)))
                }
            }
            .addOnFailureListener { done(Result.failure(googleError(it))) }
    }

    private fun googleToken(account: String, stale: String?, done: Done) {
        val ask: () -> Unit = {
            Identity.getAuthorizationClient(activity).authorize(googleRequest(account))
                .addOnSuccessListener { auth ->
                    // A screen would be needed: that is a sign-in, which only a click may start.
                    done(Result.success(if (auth.hasResolution()) JSONObject.NULL else auth.accessToken ?: JSONObject.NULL))
                }
                .addOnFailureListener { done(Result.failure(googleError(it))) }
        }
        if (stale == null) return ask()
        // Drive refused this token; Play services would hand it out again until it is cleared.
        net.execute {
            runCatching { GoogleAuthUtil.clearToken(activity, stale) }
            main.post { ask() }
        }
    }

    private fun signedIn(auth: AuthorizationResult): Any {
        val token = auth.accessToken ?: return JSONObject.NULL
        val account = auth.toGoogleSignInAccount()
        // Often empty: Play services does not always say who signed in. The page then asks Drive
        // (`about`) and keys the account by that address from then on.
        val email = account?.email ?: account?.account?.name ?: ""
        return JSONObject().put("account", email).put("label", email).put("token", token)
    }

    /**
     * Google reports a missing OAuth client either as DEVELOPER_ERROR or, from the consent screen,
     * as an internal error naming UNREGISTERED_ON_API_CONSOLE; both mean the same setup step.
     */
    private fun googleError(e: Exception): Exception =
        if (e is ApiException &&
            (e.statusCode == CommonStatusCodes.DEVELOPER_ERROR || e.message.orEmpty().contains("UNREGISTERED_ON_API_CONSOLE"))
        ) {
            IllegalStateException(
                "Google sign-in is not set up for this build of Hearth (docs/runbook/cloud-backups.md)",
            )
        } else {
            e
        }

    // ---- Dropbox -----------------------------------------------------------------------------

    private val dropboxRedirect get() = "db-${BuildConfig.DROPBOX_APP_KEY}://2/token"

    private fun dropboxSignIn(done: Done) {
        dropboxWaiting?.done?.invoke(Result.success(JSONObject.NULL))
        val verifier = randomToken(48)
        val state = randomToken(16)
        val url = Uri.parse(DROPBOX_AUTHORIZE).buildUpon()
            .appendQueryParameter("client_id", BuildConfig.DROPBOX_APP_KEY)
            .appendQueryParameter("response_type", "code")
            .appendQueryParameter("code_challenge", challenge(verifier))
            .appendQueryParameter("code_challenge_method", "S256")
            .appendQueryParameter("token_access_type", "offline")
            .appendQueryParameter("redirect_uri", dropboxRedirect)
            .appendQueryParameter("state", state)
            .build()
        dropboxWaiting = DropboxPending(verifier, state, done)
        try {
            activity.startActivity(Intent(Intent.ACTION_VIEW, url))
        } catch (_: ActivityNotFoundException) {
            dropboxWaiting = null
            done(Result.failure(IllegalStateException("no browser to sign in to Dropbox with")))
        }
    }

    /** The browser came back through the `db-<key>` scheme. False when it was not for us. */
    fun onRedirect(uri: Uri): Boolean {
        if (BuildConfig.DROPBOX_APP_KEY.isEmpty() || uri.scheme != "db-${BuildConfig.DROPBOX_APP_KEY}") return false
        val pending = dropboxWaiting ?: return true
        // Anything can open this scheme; only the answer to our own request, carrying our state,
        // is taken. PKCE means a stolen code is useless without the verifier anyway.
        if (uri.getQueryParameter("state") != pending.state) return true
        dropboxWaiting = null
        val code = uri.getQueryParameter("code")
        if (code == null) {
            // Denied on Dropbox's screen: the same as backing out.
            pending.done(Result.success(JSONObject.NULL))
            return true
        }
        background(pending.done) {
            val tokens = JSONObject(
                post(
                    DROPBOX_TOKEN,
                    form(
                        "code" to code,
                        "grant_type" to "authorization_code",
                        "code_verifier" to pending.verifier,
                        "client_id" to BuildConfig.DROPBOX_APP_KEY,
                        "redirect_uri" to dropboxRedirect,
                    ),
                ),
            )
            val access = tokens.getString("access_token")
            val account = tokens.getString("account_id")
            prefs.edit().putString(key(account), tokens.getString("refresh_token")).apply()
            remember(account, access, tokens.optLong("expires_in", 3600))
            val email = runCatching {
                JSONObject(post(DROPBOX_ACCOUNT, "null", bearer = access)).optString("email")
            }.getOrNull().orEmpty()
            JSONObject().put("account", account).put("label", email.ifEmpty { "Dropbox" }).put("token", access)
        }
        return true
    }

    /** The user came back to the app without finishing the Dropbox sign-in. */
    fun onResume() {
        val pending = dropboxWaiting ?: return
        // onNewIntent (the redirect) runs before onResume, so reaching here with the sign-in still
        // open means the browser was left some other way.
        main.postDelayed({
            if (dropboxWaiting === pending) {
                dropboxWaiting = null
                pending.done(Result.success(JSONObject.NULL))
            }
        }, RESUME_GRACE_MS)
    }

    /** On the network thread. */
    private fun dropboxToken(account: String, stale: String?): String? {
        dropboxTokens[account]?.let { (token, until) ->
            if (token != stale && System.currentTimeMillis() < until) return token
        }
        val refresh = prefs.getString(key(account), null) ?: return null
        val answer = try {
            JSONObject(
                post(
                    DROPBOX_TOKEN,
                    form(
                        "grant_type" to "refresh_token",
                        "refresh_token" to refresh,
                        "client_id" to BuildConfig.DROPBOX_APP_KEY,
                    ),
                ),
            )
        } catch (e: HttpError) {
            // 400 is Dropbox saying the grant is gone (revoked, or the app unlinked): sign in again.
            if (e.code == 400) {
                prefs.edit().remove(key(account)).apply()
                return null
            }
            throw e
        }
        val token = answer.getString("access_token")
        remember(account, token, answer.optLong("expires_in", 3600))
        return token
    }

    private fun remember(account: String, token: String, expiresIn: Long) {
        // A minute early, so a token never expires between being handed out and being used.
        dropboxTokens[account] = token to System.currentTimeMillis() + (expiresIn - 60) * 1000
    }

    // ---- plumbing ----------------------------------------------------------------------------

    private fun background(done: Done, work: () -> Any) = net.execute {
        val result = runCatching(work)
        main.post { done(result) }
    }

    private fun randomToken(bytes: Int) =
        Base64.getUrlEncoder().withoutPadding().encodeToString(ByteArray(bytes).also(random::nextBytes))

    private class DropboxPending(val verifier: String, val state: String, val done: Done)

    class HttpError(val code: Int, body: String) : IOException("Dropbox answered $code: ${body.take(200)}")

    companion object {
        private const val PREFS = "hearth-cloud"
        private const val DRIVE = "https://www.googleapis.com/auth/drive"
        private const val GOOGLE_ACCOUNT_TYPE = "com.google"
        private const val DROPBOX_AUTHORIZE = "https://www.dropbox.com/oauth2/authorize"
        private const val DROPBOX_TOKEN = "https://api.dropboxapi.com/oauth2/token"
        private const val DROPBOX_ACCOUNT = "https://api.dropboxapi.com/2/users/get_current_account"
        private const val DROPBOX_REVOKE = "https://api.dropboxapi.com/2/auth/token/revoke"
        private const val RESUME_GRACE_MS = 1500L

        private fun key(account: String) = "dropbox:$account"

        /** RFC 7636 S256: base64url of the SHA-256 of the verifier, unpadded. */
        fun challenge(verifier: String): String =
            Base64.getUrlEncoder().withoutPadding()
                .encodeToString(MessageDigest.getInstance("SHA-256").digest(verifier.toByteArray(Charsets.US_ASCII)))

        fun form(vararg fields: Pair<String, String>): String =
            fields.joinToString("&") { (k, v) -> "${URLEncoder.encode(k, "UTF-8")}=${URLEncoder.encode(v, "UTF-8")}" }

        /** A form post (or a JSON one with a bearer token), through [Egress]; the body of a 2xx, else [HttpError]. */
        private fun post(url: String, body: String, bearer: String? = null): String {
            val headers = if (bearer != null) {
                mapOf("Authorization" to "Bearer $bearer", "Content-Type" to "application/json")
            } else {
                mapOf("Content-Type" to "application/x-www-form-urlencoded")
            }
            return try {
                Egress.request("POST", url, headers, body.toByteArray()).decodeToString()
            } catch (e: Egress.HttpError) {
                throw HttpError(e.code, e.message.orEmpty())
            }
        }
    }
}

typealias Done = (Result<Any>) -> Unit
